base/floatfuncs.jl

```diff
@@ -406,22 +406,13 @@ function fma_emulated(a::Float64, b::Float64,c::Float64)
 end
 fma_llvm(x::Float32, y::Float32, z::Float32) = fma_float(x, y, z)
 fma_llvm(x::Float64, y::Float64, z::Float64) = fma_float(x, y, z)
+
 # Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
 # onto a broken system libm; if so, use a software emulated fma
-# 1.0000305f0 = 1 + 1/2^15
-# 1.0000000009313226 = 1 + 1/2^30
-# If fma_llvm() clobbers the rounding mode, the result of 0.1 + 0.2 will be 0.3
-# instead of the properly-rounded 0.30000000000000004; check after calling fma
-# TODO actually detect fma in hardware and switch on that.
-if (Sys.ARCH !== :i686 && fma_llvm(1.0000305f0, 1.0000305f0, -1.0f0) == 6.103609f-5 &&
-    (fma_llvm(1.0000000009313226, 1.0000000009313226, -1.0) ==
-     1.8626451500983188e-9) && 0.1 + 0.2 == 0.30000000000000004)
-    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
-    fma(x::Float64, y::Float64, z::Float64) = fma_llvm(x,y,z)
-else
-    fma(x::Float32, y::Float32, z::Float32) = fma_emulated(x,y,z)
-    fma(x::Float64, y::Float64, z::Float64) = fma_emulated(x,y,z)
-end
+have_fma() = ccall("extern julia.cpu.have_fma", llvmcall, Int, ()) == 1
```

Comment:
I still think this should take the type as input.

LLVM does have the option to turn off double precision for vfp4 and the support for fma for f16 and/or f128 is most likely not the same as f32 and f64.

Reply:
I would also suggest making this a `Core.Intrinsic` function

Reply:
Added a type (as if `julia.cpu.have_fma` were an overloaded intrinsic, because actually passing a jl_datatype_t and checking that from an LLVM pass was cumbersome), and made it a proper Julia intrinsic (not sure if that's worth the it though). Suggestions on how to improve both these parts are appreciated.